examples/perf/ipc.py

```diff
@@ -0,0 +1,186 @@
+from time import sleep
+import sys
+from optparse import OptionParser
+
+# load BPF program
+code="""
+#include <uapi/linux/ptrace.h>
+
+struct perf_delta {
+    u64 clk_delta;
+    u64 inst_delta;
+    u64 time_delta;
+};
+
+/*
+Perf Arrays to read counter values for open
+perf events.
+BPF_PERF_ARRAY taken maximum number of cpus.
+Change it to the max number of cpu in case
+it is greater than MAX_CPUS
+*/
+#define MAX_CPUS 128
+BPF_PERF_ARRAY(clk, MAX_CPUS);
+BPF_PERF_ARRAY(inst, MAX_CPUS);
+
+// Perf Output
+BPF_PERF_OUTPUT(output);
+
+// Per Cpu Data to store start values
+BPF_PERCPU_ARRAY(data, u64);
+
+void trace_start(struct pt_regs *ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    /*
+    perf_read may return negative values for errors.
+    If cpu id is greater than BPF_PERF_ARRAY size,
+    counters values will be very large negative number.
+    Modify MAX_CPUS to the current number of CPUs.
+    NOTE: Use bpf_perf_event_value is recommended over
+    bpf_perf_event_read or map.perf_read() due to
+    issues in ABI. map.perf_read_value() need to be
+    implemented in future.
+    */
+    u64 clk_start = clk.perf_read(cpu);
+    u64 inst_start = inst.perf_read(cpu);
+    u64 time_start = bpf_ktime_get_ns();
+    
+    u64* kptr = NULL;
+    kptr = data.lookup(&clk_k);
+    if (kptr) {
+        data.update(&clk_k, &clk_start);
+    } else {
+        data.insert(&clk_k, &clk_start);
+    }
+
+    kptr = data.lookup(&inst_k);
+    if (kptr) {
+        data.update(&inst_k, &inst_start);
+    } else {
+        data.insert(&inst_k, &inst_start);
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        data.update(&time, &time_start);
+    } else {
+        data.insert(&time, &time_start);
+    }
+}
+
+void trace_end(struct pt_regs* ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    /*
+    perf_read may return negative values for errors.
+    If cpu id is greater than BPF_PERF_ARRAY size,
+    counters values will be very large negative number.
+    Modify MAX_CPUS to the current number of CPUs.
+    NOTE: Use bpf_perf_event_value is recommended over
+    bpf_perf_event_read or map.perf_read() due to
+    issues in ABI. map.perf_read_value() need to be
+    implemented in future.
+    */
+    u64 clk_end = clk.perf_read(cpu);
+    u64 inst_end = inst.perf_read(cpu);
+    u64 time_end = bpf_ktime_get_ns();
+    
+    struct perf_delta perf_data = {} ;
+    bool submit = true;
+    u64* kptr = NULL;
+    kptr = data.lookup(&clk_k);
+    if (kptr) {
+        perf_data.clk_delta = clk_end - *kptr;
+    }
+    
+    kptr = data.lookup(&inst_k);
+    if (kptr) {
+        perf_data.inst_delta = inst_end - *kptr;
+    } else {
+        submit = false;
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        perf_data.time_delta = time_end - *kptr;
+    } else {
+        submit = false;
```

Comment:
Here and above maybe we could short-circuit this with a `return`

Reply:
Removed submit, and put a return in case of map lookups fail.